firedrake/function.py

```diff
@@ -294,16 +294,17 @@ def interpolate(self, expression, subset=None):
 
         :param expression: :class:`.Expression` to interpolate
         :returns: this :class:`Function` object"""
+        assert isinstance(expression, ufl.classes.Expr)
 
         # Make sure we have an expression of the right length i.e. a value for
         # each component in the value shape of each function space
         dims = [np.prod(fs.ufl_element().value_shape(), dtype=int)
                 for fs in self.function_space()]
-        if np.prod(expression.value_shape(), dtype=int) != sum(dims):
+        if np.prod(expression.ufl_shape, dtype=int) != sum(dims):
             raise RuntimeError('Expression of length %d required, got length %d'
-                               % (sum(dims), np.prod(expression.value_shape(), dtype=int)))
+                               % (sum(dims), np.prod(expression.ufl_shape, dtype=int)))
 
-        if hasattr(expression, 'eval'):
+        if not isinstance(expression, expression_t.Expression) or hasattr(expression, 'eval'):
             fs = self.function_space()
             if isinstance(fs, functionspace.MixedFunctionSpace):
                 raise NotImplementedError(
```

Comment:
I don't claim to support interpolating onto mixed functions, although one can always interpolate onto `f.sub(i)`. Though I haven't tried it, it might just work. Interpolation from UFL expressions containing mixed functions, however, works and is tested.